[zm] New Zoneminder binding initial contribution

mhilbush opened the pull request:
The current OH2 Zoneminder binding (`org.openhab.binding.zoneminder`) is unmaintained. 
It also has a number of shortcomings:
- There are 6266 SLOC (as counted by SLOCCount). This count includes the zoneminder binding itself, as well as the zoneminder4j library that the binding uses.
- The binding was written for older versions of Zoneminder. Since then, the Zoneminder API has evolved considerably, such that much of this binding's functionality can now be accomplished much more easily with version 2.0 of the Zoneminder API.
- The current binding doesn't support Zoneminder's latest JWT-based authentication scheme, which improves security over the previous authentication methods. Anyone who has enabled authentication in recent versions of Zoneminder is unable to use the binding.
- The current binding (and the library it uses) do not use null annotations. Adding null annotations is a significant amount of work.
- There is an abandoned 2-year old PR #3583 that added 4429 LOC and deleted 1649 LOC. That PR was closed a year ago. I took a look at the PR's review comments, as well as the code. Resurrecting it would be a monumental amount of work.

I wrote a new version of the binding using Zoneminder version 1.34.0 and Zoneminder API version 2.0 as a base. My binding is 1698 SLOC (as counted by SLOCCount) versus the current binding's 6266 SLOC. My original plan was for this binding to be just for my own use, and not submit it for inclusion into the distribution. But now there are a number of people using it. 🙄  See community post here (https://community.openhab.org/t/new-zoneminder-binding-for-zoneminder-versions-1-34-0/95112).

To clearly differentiate my binding from the one in the distribution, I named my binding `org.openhab.binding.zm`, which is a commonly-used shorthand for Zoneminder.

I'm looking for some guidance from @openhab/add-ons-maintainers on how to proceed.
A couple points to consider:
- I've looked at the current version of the binding, and the now-closed PR. The amount of work to bring this back to life, and make it forward-compatible with recent Zoneminder versions, is not worth my time.
- If we replace the current version of the binding in the distribution with the version I wrote, there likely will be some people who are running old versions of Zoneminder. If they upgrade to the latest version of openHAB, their Zoneminder integration will break.

I would propose the following:
- Keep the `org.openhab.binding.zm` name for 2.5.x.
- For 3.0, remove the old `org.openhab.binding.zoneminder` binding. If it's desirable to maintain the old binding name, we can rename `org.openhab.binding.zm` to `org.openhab.binding.zoneminder` at that time.

Thoughts?

Signed-off-by: Mark Hilbush <mark@hilbush.com>

mhilbush commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/action/ZmActions.java:
```diff
@@ -0,0 +1,133 @@
+ */
+@ThingActionsScope(name = "zm")
+@NonNullByDefault
+public class ZmActions implements ThingActions {
```
Good catch. Thanks.

mhilbush commented:
As there are no dissenting opinions to my proposal, I'm going to remove the WIP.

openhab-5iver commented:
I think this should replace the old addon and keep it's name. I see this as rewrite and upgrade of the existing binding. Either way, the old zoneminder binding served it's purpose well for a long time, but it should be removed (thank you, @Mr-Eskildsen!).

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,648 @@
+        int timeout = (localRefreshInterval * 1000) - 500;
+        Request request = httpClient.newRequest(buildStreamUrl(id, STREAM_IMAGE));
+        request.method(HttpMethod.GET);
+        request.timeout(timeout, TimeUnit.MILLISECONDS);
```
What happens if the user configures a refresh interval of 10min? You could set an upper limit to the timeout.

mhilbush replied:
Good idea.

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,648 @@
+                            .setStateOptions(new ChannelUID(getThing().getUID(), CHANNEL_VIDEO_MONITOR_ID), options);
+                }
+                // Only update alarm and event info for monitors whose handlers are initialized
+                Set<String> ids = monitorHandlers.keySet();
+                for (Monitor m : monitorList) {
+                    if (ids.contains(m.getId())) {
+                        m.setState(getState(m.getId()));
+                        m.setLastEvent(getLastEvent(m.getId()));
+                    }
+                }
```
The key set is backed by the map and if I see correctly, access is not synchronized. Do you need additional synchronization here?

mhilbush replied:
I see what you mean. But the Map is defined as a `ConcurrentHashMap`. Is there additional synchronization that's needed? I so, can you suggest how to do that?

fwolter replied:
`ConcurrentHashMap.keySet()` is *weakly consistent*. That means the view to the keys can change during iteration. So, you might miss an entry. But if I read the context correctly, this shouldn't be a problem and you won't need to take any action. But please double check, since this would be a functional issue.

mhilbush replied:
Thanks for the explanation. `monitorHandlers` will change only when monitor things are added or removed (a very infrequent occurrence other than during initial setup). Nonetheless, would the correct solution be to create a synchronized block on `monitorHandlers`?

fwolter replied:
Yes, and to all other accesses to the map. To make that easier, you could instead create a synchronized map via `Collections.synchronizedMap(new HashMap<>())`. Then, synchronized blocks only need to be added to loops. But as I said, I think it's ok to go with the current implementation. Only if you want to be really save...

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,648 @@
+                return true;
+            }
+        } else {
+            logger.warn("Bridge: Can't get version information from server");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Can't get version information");
```
Can the log message be removed? See point 5 https://www.openhab.org/docs/developer/guidelines.html#f-logging

mhilbush replied:
I found this to be useful when debugging user issues, so I would prefer to keep. I downgraded it to INFO level.

fwolter replied:
The custom configuration error message, supplied to `updateStatus()` is logged by the framework. The user simply has to increase the log level. I think there is no need for logging the same message to different loggers.

mhilbush replied:
Ok

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmMonitorHandler.java:
```diff
@@ -0,0 +1,328 @@
+    }
+
+    @Override
+    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
```
Is this method necessary? Child things are updated by the framework if the bridge becomes offline.

mhilbush replied:
IDK. I thought this was necessary in order for the child handlers to be informed when the bridge status changes. It is used pretty widely across many bindings.

fwolter replied:
Maybe I miss something. @cpmeister Is there any reason why the bridge should update its child states to offline when going offline itself?

fwolter replied:
I tested it: Things are automatically set offline, when the bridge goes offline. So, this method could be removed.

mhilbush replied:
Changed

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmMonitorHandler.java:
```diff
@@ -0,0 +1,328 @@
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.unmodifiableList(Stream.of(ZmActions.class).collect(Collectors.toList()));
```
```suggestion
        return Collections.singleton(ZmActions.class);
```

mhilbush replied:
So much simpler!

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmMonitorHandler.java:
```diff
@@ -0,0 +1,328 @@
+        }
+    }
+
+    private void updateMonitorStatus() {
```
See above

mhilbush replied:
This is called from `initialize`, so I think it's needed to properly set the handler status upon completion of `initialize`.

fwolter replied:
Ok, if updating the child handlers is not necessary, I think this could be replaced by a one-liner, setting the state to online.

fwolter replied:
With above insight, the whole method could replaced by `updateStatus(ThingStatus.ONLINE)`.

mhilbush replied:
Changed

fwolter commented on bundles/org.openhab.binding.zm/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,86 @@
+		<parameter name="host" type="text" required="true" groupName="url-info">
+			<label>Server</label>
+			<description>Zoneminder server name or IP address</description>
+			<context>network</context>
```
```suggestion
			<context>network-address</context>
```

mhilbush replied:
Good catch!

fwolter commented on bundles/org.openhab.binding.zm/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,86 @@
+			<label>Monitor Id</label>
+			<description>Id of the monitor</description>
+		</parameter>
+		<parameter name="imageRefreshInterval" type="integer" required="false">
```
```suggestion
		<parameter name="imageRefreshInterval" type="integer" unit="s" required="false">
```

mhilbush replied:
Done.

fwolter commented on bundles/org.openhab.binding.zm/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,86 @@
+			<label>Image Refresh Interval</label>
+			<description>Interval in seconds with which monitor image is refreshed</description>
+		</parameter>
+		<parameter name="alarmDuration" type="integer" required="false">
```
```suggestion
		<parameter name="alarmDuration" type="integer" unit="s" required="false">
```

mhilbush replied:
Done.

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,648 @@
+                errorMsg = String.format("HTTP GET failed: %d, %s", response.getStatus(), response.getReason());
+            }
+        } catch (TimeoutException e) {
+            errorMsg = String.format("TimeoutException: Call to Doorbird API timed out after {} msec", timeout);
```
Doorbird?

mhilbush replied:
Good spot. Copy/paste error. 🤭

mhilbush commented:
> I think this should replace the old addon and keep it's name. 

> IMHO, it would be a good way to replace the old binding by this one, when OH3 is released, as you suggested. But I would keep the name zoneminder.

@openhab-5iver @fwolter So, if I understand you both, you agree with leaving the binding named `zm` for 2.5.x, but for 3.0 rename it `zoneminder` and remove the old binding.

To rename it for 2.5.x, it would badly break anyone who's using the old binding. In the event that those users are using something less than Zoneminder 1.35.0, they would be completely screwed, as they may not be in a position to upgrade their Zoneminder installation..

mhilbush commented on bundles/org.openhab.binding.zm/README.md:
```diff
@@ -0,0 +1,276 @@
+|-----------|--------------|-------------------|-------------|
+| Host                           | host                        | Required  | Host name or IP address of the Zoneminder server. |
+| Port Number                    | portNumber                  | Optional  | Port number if not on Zoneminder default port 80. |
+| Use Default Url Path           | useDefaultUrlPath           | Required  | If ON, use Zoneminder default of /zm; if OFF, use value from urlPath parameter. |
+| Url Path                       | urlPath                     | Optional  | Use this parameter when not using the Zoneminder default path of /zm. |
```
> Add "/zm" as default value to the urlPath parameter and remove the switch parameter

This is how I wanted to do it originally. However, if you define the parameter to have a default, I don't believe it's then possible to leave the parameter empty. If you leave it empty, OH will just insert the default. As there are some Zoneminder installations that need the path to be empty, this was the only way I could think to do it.

> Did you think about letting the user paste the complete URL in one configuration parameter?

I did not.

mhilbush commented on bundles/org.openhab.binding.zm/README.md:
```diff
@@ -0,0 +1,276 @@
+
+### Requirements
+
+The binding requires Zoneminder version 1.34.0 or greater, and API version 2.0 or greater.
+The API must be enabled in the Zoneminder configuration using the **OPT_USE_API** parameter.
+
+The binding can access Zoneminder with or without authentication.
+If you have authentication enabled in Zoneminder, set the **user** and **pass** 
+config paremeters in the *server* thing.
```
Understood. I was trying to address several users who had questions about auth versus non-auth Zoneminder configs. I wanted to explicitly cover what was required when the user has Zoneminder auth enabled.

mhilbush commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/action/ZmActions.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    public static void triggerAlarm(@Nullable ThingActions actions, @Nullable Number alarmDuration) {
+        invokeMethodOf(actions).triggerAlarm(alarmDuration);
```
Thanks. I was just about to make a reference to the issue.

mhilbush commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/discovery/MonitorDiscoveryService.java:
```diff
@@ -0,0 +1,140 @@
+        return bridgeHandler;
+    }
+
+    @Override
+    public void activate() {
+        logger.debug("Discovery: Activating Zoneminder monitor discovery service for {}", getBridgeUID());
+    }
```
Yes, it was helpful during development, but no longer needed.

mhilbush commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/discovery/MonitorDiscoveryService.java:
```diff
@@ -0,0 +1,140 @@
+                Integer imageRefreshInterval = localBridgeHandler.getDefaultImageRefreshInterval();
+                ThingUID bridgeUID = localBridgeHandler.getThing().getUID();
+                ThingUID thingUID = new ThingUID(UID_MONITOR, monitor.getId());
+                Map<String, Object> properties = new HashMap<>(0);
```
No. Fixed.

mhilbush commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/discovery/MonitorDiscoveryService.java:
```diff
@@ -0,0 +1,140 @@
+
+    private DiscoveryResult createDiscoveryResult(ThingUID monitorUID, ThingUID bridgeUID, String id, String name,
+            Map<String, Object> properties) {
+        return DiscoveryResultBuilder.create(monitorUID).withProperties(properties).withBridge(bridgeUID)
+                .withLabel(buildLabel(name)).build();
```
I had hoped to be able to do this. But, the only thing unique about a monitor is its ID, which is an integer value that in almost all cases will be a single digit. I wasn't sure if that really was sufficient for a representation property? WDYT?

fwolter replied:
I think, if it's unique, it doesn't matter how long the identifier is. I would go for it.

mhilbush replied:
Ok, done.

fwolter replied:
Did you push your changes? I don't see the representation property.

mhilbush replied:
I haven't pushed the changes yet. I'll do so after making a few additional changes.

mhilbush commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,648 @@
+
+    @Override
+    public void initialize() {
+        logger.debug("Bridge: Handler initializing");
```
Agreed. Useful during development but no longer needed.

mhilbush commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,648 @@
+    @Override
+    public void dispose() {
+        cancelRefreshJob();
+        logger.debug("Bridge: Handler disposing");
```
Agreed. Useful during development but no longer needed.

mhilbush commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/discovery/MonitorDiscoveryService.java:
```diff
@@ -0,0 +1,140 @@
+    @Override
+    protected synchronized void stopScan() {
+        super.stopScan();
+    }
```
Removed.

mhilbush commented:
@fwolter Thanks for your review!

I pushed changes that address most of your review comments. For the other ones, I've left answers and/or follow-up questions.

openhab-5iver commented:
My understanding is that there are a number of people using the old binding with an older version of ZM and that they have been stuck with those versions, without the ability to upgrade. There are also many people who are using a newer version of ZM and who were not able to hack at things to get the older version of the binding working. Many have dropped zoneminder due to lack of support in OH. IMO, it is more important to be compatible with the current version of ZM. I don't see this as a breaking change to the ZM binding, but a response to the many breaking changes in ZM.

It is my understanding that the current version of the binding distributed with OH does not work with ZM 1.30 and newer, and that pretty much everyone is using the beta version from the marketplace. So, people will still be able to download and install it from there, and upgrading the one in the OH distro will not affect them. I think the one in the marketplace also works with older versions of ZM. A poll in the forum would help to understand the impact to the community.

mhilbush commented:
> A poll in the forum would help to understand the impact to the community.

Only helpful for those who follow the forum consistently, and are inclined to respond to polls. 

Unfortunately, there's no real way to know about binding usage. I'm just trying to be the least disruptive in the short term. OH3 will be here soon enough. That's the opportunity to be more disruptive.

Push come to shove, I'll just not add this to 2.5.x

mhilbush commented:
@fwolter I believe I've addressed all the outstanding review feedback. Apologize for the delay as I've been sidetracked on a few other things.

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,645 @@
+        setAlarm(buildUrl(String.format("/api/monitors/alarm/id:%s/command:off.json", id)));
+    }
+
+    public @Nullable RawType getImage(String id, @Nullable Integer imageRefreshInterval) {
```
It's good practice to append the unit to the name e.g. TIMEOUT_SEC.

mhilbush replied:
Done

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmMonitorHandler.java:
```diff
@@ -0,0 +1,293 @@
+    private @Nullable ZmBridgeHandler bridgeHandler;
+
+    private @NonNullByDefault({}) String monitorId;
+    private @Nullable Integer imageRefreshInterval;
+    private Integer alarmDuration = DEFAULT_ALARM_DURATION_SECONDS;
```
It's good practice to append the unit to the field name e.g. alarmDurationSec.

mhilbush replied:
Done

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,645 @@
+                    System.currentTimeMillis() - startTime);
+            return response;
+        } catch (IOException e) {
+            logger.debug("Bridge: IOException on url '{}': {}", url, e.getMessage(), e);
```
The stack trace should only be logged if something unexpected happened, like the detection of a bug in your code. Same for below. A plain log message should be enough when the network fails.
```suggestion
            logger.debug("Bridge: IOException on url '{}': {}", url, e.getMessage());
```

mhilbush replied:
Fixed

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,645 @@
+                        logger.debug("Bridge: Zoneminder software version check OK");
+                        return true;
+                    } else {
+                        logger.warn("Bridge: Requires version >= 1.34.0. This Zoneminder is version {}",
+                                softwareVersion);
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                                "Requires version >= 1.34.0");
```
This is already logged by the framework via `updateStatus()`. You could remove the log message. Same for below.

mhilbush replied:
All logs removed

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,645 @@
+        try {
+            refreshMonitors();
+            discoverMonitors();
+        } catch (RuntimeException e) {
```
What type of exception do you expect here? Can you specify the concrete type?

mhilbush replied:
I'm trying to prevent the refresh job from silently failing due to an unexpected exception. The exception normally would result from a poorly formatted response from the Zoneminder server, but the exact exception type is hard to predict. If not caught, the refresh job will be canceled and nothing will be logged. Catching the runtime exception is here to prevent that situation.

fwolter replied:
You already catch the JsonSyntaxException and IOExceptions. That should cover most unexpected cases. However, uncatched unchecked exceptions like `RuntimeException` are logged by the framework. They don't need to be logged by the binding.

mhilbush replied:
Understood. However, I'm pretty sure I added this because I got an exception (I don't remember which one) that caused the refresh job to be canceled and was not logged by the framework. I can remove this, then revisit if the problem reoccurs.

fwolter replied:
If it wasn't logged, it might be a bug in the framework. Would be good to remove it and investigate that.

mhilbush replied:
Ok, I removed it here and one other place.

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmBridgeHandler.java:
```diff
@@ -0,0 +1,645 @@
+            refreshMonitors();
+            discoverMonitors();
+        } catch (RuntimeException e) {
+            logger.info("Bridge: Unexpected exception during refresh", e);
```
Logging to info should be used rarely. This could be warn.

mhilbush replied:
Agreed. However, in this case since the exception is not expected, and may not be easily reproduced, I would prefer it be logged when it occurs. Ok?

mhilbush replied:
I'll make it warn.

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmMonitorHandler.java:
```diff
@@ -0,0 +1,293 @@
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            // No need to handle as status will be updated at the next refresh interval
```
Can you trigger an immediate refresh?

mhilbush replied:
In another binding I cached the channel states, then pulled from the cache when a refresh is received for the channel. I'll do the same thing here.

fwolter replied:
Caching should only be done if an immediate poll is not feasible, e.g. the device sends only event-based and has no polling support. Calling the poll job would be the preferred solution here.

mhilbush replied:
Understood. But I don't want to poll for each channel that gets a REFRESH command. At startup, wouldn't that result in multiple polls of the server for each monitor, which clearly is not desirable. Do you have a suggestion for how to approach this in a way that won't cause multiple server polls?

fwolter replied:
Many bindings face this challenge. For this use case [ExpiringCache ](https://www.openhab.org/javadoc/v2.5/org/eclipse/smarthome/core/cache/expiringcache) was introduced.

fwolter replied:
When I read our conversation again, you might have implemented exactly this. This should be a cache with a few seconds expiring time.

mhilbush replied:
Thanks. I did something similar along the lines of what I did in the weathercompany binding. It was an easy change, and would require less rework than adapting to use ExpiringCache. I'll include it in what I push, then you can weigh in on whether it's acceptable.

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmMonitorHandler.java:
```diff
@@ -0,0 +1,293 @@
+        logger.debug("Monitor {}: Received command '{}' for channel '{}'", monitorId, command, channelUID.getId());
+        ZmBridgeHandler localHandler = bridgeHandler;
+        if (localHandler == null) {
+            logger.info("Monitor {}: Can't execute command because bridge handler is null", monitorId);
```
See above

mhilbush replied:
Implemented cache of channel states to enable handling of REFRESH.

fwolter commented on bundles/org.openhab.binding.zm/src/main/java/org/openhab/binding/zm/internal/handler/ZmMonitorHandler.java:
```diff
@@ -0,0 +1,293 @@
+                RawType image = localHandler.getImage(monitorId, imageRefreshInterval);
+                updateState(CHANNEL_IMAGE, image != null ? image : UnDefType.UNDEF);
+            }
+        } catch (RuntimeException e) {
```
See above

mhilbush replied:
As mentioned above I'm trying to avoid a poorly formatted response from Zoneminder from causing an unexpected exception that causes the image refresh job to be canceled silently.

mhilbush replied:
Removed

fwolter commented on bundles/org.openhab.binding.zm/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,86 @@
+			<description>Zoneminder server name or IP address</description>
+			<context>network-address</context>
+		</parameter>
+		<parameter name="portNumber" type="text" required="false" groupName="url-info">
```
You might want to specify min/max values.
```suggestion
		<parameter name="portNumber" type="integer" required="false" groupName="url-info">
```

mhilbush replied:
Good idea. Done.

fwolter commented on bundles/org.openhab.binding.zm/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,86 @@
+	</config-description>
+
+	<config-description uri="thing-type:zm:monitor">
+		<parameter name="monitorId" type="text" required="true">
```
Is this really text or integer?

mhilbush replied:
Zoneminder treats it as text, so I wanted to be consistent. Even though they appear to be all numeric now, I don't think it can be guaranteed to stay that way.

mhilbush commented:
@fwolter Sorry for leaving this sit for so long. Real life has a habit of getting in the way sometimes. Thanks for your patience!

I'm working on addressing your feedback and will ping you once I push the changes.

mhilbush commented:
@fwolter I've pushed my changes that address your review feedback.

In addition to those changes, I've also addressed several issues that users have reported on the forum:
- url encode authentication string to handle special characters in user name and password
- add config parameter to select http or https for zoneminder connection

While you're looking over my changes, I'm going to move my development over to OH3. In the process of doing that, I'm going to rename the binding from `zm` to `zoneminder` in order to replace the old and unsupported zoneminder binding.

mhilbush commented:
I'm closing this in favor of opening a new PR for OH3.
